File: airbyte-integrations/connectors/destination-redshift/src/test-integration/java/io/airbyte/integrations/destination/redshift/RedshiftTestDataComparator.java
/*
 * Copyright (c) 2022 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.integrations.destination.redshift;

import io.airbyte.integrations.standardtest.destination.comparator.AdvancedTestDataComparator;
import java.time.LocalDate;
import java.time.ZoneOffset;
import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;
import java.time.format.DateTimeParseException;
import java.util.ArrayList;
import java.util.List;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class RedshiftTestDataComparator extends AdvancedTestDataComparator {

  private static final Logger LOGGER = LoggerFactory.getLogger(RedshiftTestDataComparator.class);

  private final RedshiftSQLNameTransformer namingResolver = new RedshiftSQLNameTransformer();

  protected static final String REDSHIFT_DATETIME_WITH_TZ_FORMAT = "yyyy-MM-dd HH:mm:ssX";

  @Override
  protected List<String> resolveIdentifier(final String identifier) {
    final List<String> result = new ArrayList<>();
    final String resolved = namingResolver.getIdentifier(identifier);
    result.add(identifier);
    result.add(resolved);
    if (!resolved.startsWith("\"")) {
      result.add(resolved.toLowerCase());
      result.add(resolved.toUpperCase());
    }
    return result;
  }

  @Override
  protected boolean compareDateTimeWithTzValues(final String airbyteMessageValue,
                                                final String destinationValue) {
    try {
      final ZonedDateTime airbyteDate = ZonedDateTime.parse(airbyteMessageValue,
          getAirbyteDateTimeWithTzFormatter()).withZoneSameInstant(
              ZoneOffset.UTC);

      final ZonedDateTime destinationDate = ZonedDateTime.parse(destinationValue,
          DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ssX"));
      return airbyteDate.equals(destinationDate);
    } catch (DateTimeParseException e) {
      LOGGER.warn(
          "Fail to convert values to ZonedDateTime. Try to compare as text. Airbyte value({}), Destination value ({}). Exception: {}",
          airbyteMessageValue, destinationValue, e);
      return compareTextValues(airbyteMessageValue, destinationValue);
    }
  }

  @Override
  protected boolean compareDateTimeValues(String expectedValue, String actualValue) {
    var destinationDate = parseLocalDateTime(actualValue);
    var expectedDate = LocalDate.parse(expectedValue,
        DateTimeFormatter.ofPattern(AIRBYTE_DATETIME_FORMAT));
    return expectedDate.equals(destinationDate);
  }

  private LocalDate parseLocalDateTime(String dateTimeValue) {
    if (dateTimeValue != null) {
      return LocalDate.parse(dateTimeValue,
          DateTimeFormatter.ofPattern(getFormat(dateTimeValue)));
    } else {
      return null;
    }
  }

  private String getFormat(String dateTimeValue) {
    if (dateTimeValue.contains("T")) {
      // MySql stores array of objects as a jsonb type, i.e. array of string for all cases
      return AIRBYTE_DATETIME_FORMAT;
    } else {
      // MySql stores datetime as datetime type after normalization
      return AIRBYTE_DATETIME_PARSED_FORMAT;
    }
  }

}

File: airbyte-integrations/connectors/destination-redshift/src/test-integration/kotlin/io/airbyte/integrations/destination/redshift/RedshiftTestDataComparator.kt
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */
package io.airbyte.integrations.destination.redshift

import edu.umd.cs.findbugs.annotations.SuppressFBWarnings
import io.airbyte.cdk.integrations.standardtest.destination.comparator.AdvancedTestDataComparator
import java.time.LocalDate
import java.time.ZoneOffset
import java.time.ZonedDateTime
import java.time.format.DateTimeFormatter
import java.time.format.DateTimeParseException
import java.util.Locale
import org.slf4j.Logger
import org.slf4j.LoggerFactory

class RedshiftTestDataComparator : AdvancedTestDataComparator() {
    private val namingResolver = RedshiftSQLNameTransformer()

    @SuppressFBWarnings("NP_PARAMETER_MUST_BE_NONNULL_BUT_MARKED_AS_NULLABLE")
    override fun resolveIdentifier(identifier: String?): List<String?> {
        val result: MutableList<String?> = ArrayList()
        val resolved = namingResolver.getIdentifier(identifier!!)
        result.add(identifier)
        result.add(resolved)
        if (!resolved.startsWith("\"")) {
            result.add(resolved.lowercase(Locale.getDefault()))
            result.add(resolved.uppercase(Locale.getDefault()))
        }
        return result
    }

    override fun compareDateTimeWithTzValues(
        airbyteMessageValue: String,
        destinationValue: String
    ): Boolean {
        try {
            val airbyteDate =
                ZonedDateTime.parse(airbyteMessageValue, airbyteDateTimeWithTzFormatter)
                    .withZoneSameInstant(ZoneOffset.UTC)

            val destinationDate =
                ZonedDateTime.parse(destinationValue).withZoneSameInstant(ZoneOffset.UTC)
            return airbyteDate == destinationDate
        } catch (e: DateTimeParseException) {
            LOGGER.warn(
                "Fail to convert values to ZonedDateTime. Try to compare as text. Airbyte value({}), Destination value ({}). Exception: {}",
                airbyteMessageValue,
                destinationValue,
                e
            )
            return compareTextValues(airbyteMessageValue, destinationValue)
        }
    }

    override fun compareDateTimeValues(expectedValue: String, actualValue: String): Boolean {
        val destinationDate = parseLocalDateTime(actualValue)
        val expectedDate =
            LocalDate.parse(expectedValue, DateTimeFormatter.ofPattern(AIRBYTE_DATETIME_FORMAT))
        return expectedDate == destinationDate
    }

    private fun parseLocalDateTime(dateTimeValue: String?): LocalDate? {
        return if (dateTimeValue != null) {
            LocalDate.parse(dateTimeValue, DateTimeFormatter.ofPattern(getFormat(dateTimeValue)))
        } else {
            null
        }
    }

    private fun getFormat(dateTimeValue: String): String {
        return if (dateTimeValue.contains("T")) {
            // MySql stores array of objects as a jsonb type, i.e. array of string for all cases
            AIRBYTE_DATETIME_FORMAT
        } else {
            // MySql stores datetime as datetime type after normalization
            AIRBYTE_DATETIME_PARSED_FORMAT
        }
    }

    companion object {
        private val LOGGER: Logger = LoggerFactory.getLogger(RedshiftTestDataComparator::class.java)

        protected const val REDSHIFT_DATETIME_WITH_TZ_FORMAT: String = "yyyy-MM-dd HH:mm:ssX"
    }
}
